File: packages/sdk/clawdstrike-hunt/src/correlate/ioc.ts
import * as fs from "node:fs/promises";
import { IocError } from "../errors.js";
import type { IocEntry, IocMatch, IocType, TimelineEvent } from "../types.js";

// ---------------------------------------------------------------------------
// Auto-detection
// ---------------------------------------------------------------------------

/**
 * Auto-detect the IOC type from the indicator string format.
 */
export function detectIocType(indicator: string): IocType | undefined {
  const trimmed = indicator.trim();
  if (trimmed.length === 0) return undefined;

  const lower = trimmed.toLowerCase();

  // Hex-only hashes
  if (lower.length === 64 && /^[0-9a-f]+$/.test(lower)) return "sha256";
  if (lower.length === 40 && /^[0-9a-f]+$/.test(lower)) return "sha1";
  if (lower.length === 32 && /^[0-9a-f]+$/.test(lower)) return "md5";

  // URL
  if (lower.startsWith("http://") || lower.startsWith("https://")) return "url";

  // IPv4
  if (isIpv4(trimmed)) return "ipv4";

  // IPv6: contains colons and hex digits
  if (trimmed.includes(":") && /^[0-9a-fA-F:]+$/.test(trimmed)) return "ipv6";

  // Domain: contains dot, no spaces, no slashes, no colons
  if (
    trimmed.includes(".") &&
    !trimmed.includes(" ") &&
    !trimmed.includes("/") &&
    !trimmed.includes(":")
  ) {
    return "domain";
  }

  return undefined;
}

function isIpv4(s: string): boolean {
  const parts = s.split(".");
  if (parts.length !== 4) return false;
  return parts.every((p) => {
    if (p.length === 0) return false;
    const n = Number(p);
    return Number.isInteger(n) && n >= 0 && n <= 255 && String(n) === p;
  });
}

// ---------------------------------------------------------------------------
// Word-boundary matching
// ---------------------------------------------------------------------------

function isIocWordChar(ch: number): boolean {
  // alphanumeric, dot, hyphen
  return (
    (ch >= 0x30 && ch <= 0x39) || // 0-9
    (ch >= 0x41 && ch <= 0x5a) || // A-Z
    (ch >= 0x61 && ch <= 0x7a) || // a-z
    ch === 0x2e || // .
    ch === 0x2d    // -
  );
}

/**
 * Check whether `needle` appears in `haystack` at word boundaries.
 * Word chars = alphanumeric + '.' + '-'.
 */
export function containsWordBounded(haystack: string, needle: string): boolean {
  if (needle.length === 0) return false;

  let start = 0;
  while (true) {
    const pos = haystack.indexOf(needle, start);
    if (pos === -1) return false;

    const endPos = pos + needle.length;
    const leftOk = pos === 0 || !isIocWordChar(haystack.charCodeAt(pos - 1));
    const rightOk = endPos >= haystack.length || !isIocWordChar(haystack.charCodeAt(endPos));

    if (leftOk && rightOk) return true;
    start = pos + 1;
  }
}

// ---------------------------------------------------------------------------
// CSV parsing
// ---------------------------------------------------------------------------

function splitCsvFields(line: string): string[] {
  const fields: string[] = [];
  let current = "";
  let inQuotes = false;
  let i = 0;

  while (i < line.length) {
    const ch = line[i];
    if (inQuotes) {
      if (ch === '"') {
        if (i + 1 < line.length && line[i + 1] === '"') {
          current += '"';
          i += 2;
          continue;
        }
        inQuotes = false;
      } else {
        current += ch;
      }
    } else if (ch === '"') {
      inQuotes = true;
    } else if (ch === ",") {
      fields.push(current.trim());
      current = "";
    } else {
      current += ch;
    }
    i++;
  }
  fields.push(current.trim());
  return fields;
}

function parseIocTypeStr(s: string): IocType | undefined {
  switch (s.trim().toLowerCase()) {
    case "sha256":
    case "sha-256":
      return "sha256";
    case "sha1":
    case "sha-1":
      return "sha1";
    case "md5":
      return "md5";
    case "domain":
    case "domain-name":
      return "domain";
    case "ipv4":
    case "ipv4-addr":
    case "ip":
      return "ipv4";
    case "ipv6":
    case "ipv6-addr":
      return "ipv6";
    case "url":
      return "url";
    default:
      return undefined;
  }
}

function parseCsvLine(line: string): IocEntry | undefined {
  const fields = splitCsvFields(line);
  if (fields.length === 0) return undefined;

  const indicator = fields[0].trim();
  if (indicator.length === 0) return undefined;

  let iocType: IocType | undefined;
  if (fields.length > 1 && fields[1].length > 0) {
    iocType = parseIocTypeStr(fields[1]) ?? detectIocType(indicator);
  } else {
    iocType = detectIocType(indicator);
  }

  if (iocType === undefined) return undefined;

  return {
    indicator,
    iocType,
    description: fields[2] && fields[2].length > 0 ? fields[2] : undefined,
    source: fields[3] && fields[3].length > 0 ? fields[3] : undefined,
  };
}

// ---------------------------------------------------------------------------
// STIX pattern parsing
// ---------------------------------------------------------------------------

function stixLhsToIocType(lhs: string): IocType | undefined {
  const lower = lhs.toLowerCase();
  if (lower.includes("sha-256") || lower.includes("sha256")) return "sha256";
  if (lower.includes("sha-1") || lower.includes("sha1")) return "sha1";
  if (lower.includes("md5")) return "md5";
  if (lower.startsWith("domain-name")) return "domain";
  if (lower.startsWith("ipv4-addr")) return "ipv4";
  if (lower.startsWith("ipv6-addr")) return "ipv6";
  if (lower.startsWith("url")) return "url";
  return undefined;
}

function parseStixPattern(pattern: string): { indicator: string; iocType: IocType } | undefined {
  const trimmed = pattern.trim();
  if (!trimmed.startsWith("[") || !trimmed.endsWith("]")) return undefined;

  const inner = trimmed.slice(1, -1);
  const eqIdx = inner.indexOf("=");
  if (eqIdx === -1) return undefined;

  const lhs = inner.slice(0, eqIdx).trim();
  const rhs = inner.slice(eqIdx + 1).trim();

  // Extract value from single quotes
  if (!rhs.startsWith("'") || !rhs.endsWith("'")) return undefined;
  const value = rhs.slice(1, -1);
  if (value.length === 0) return undefined;

  const iocType = stixLhsToIocType(lhs);
  if (iocType === undefined) return undefined;

  return { indicator: value, iocType };
}

// ---------------------------------------------------------------------------
// IocDatabase
// ---------------------------------------------------------------------------

/**
 * In-memory IOC database with index structures for fast lookup.
 */
export class IocDatabase {
  private entries: IocEntry[] = [];
  private hashIndex: Map<string, number[]> = new Map();
  private domainIndex: Map<string, number[]> = new Map();
  private ipIndex: Map<string, number[]> = new Map();
  private urlIndex: Map<string, number[]> = new Map();

  constructor() {}

  addEntry(entry: IocEntry): void {
    const indicator = entry.indicator.trim();
    if (indicator.length === 0) return;

    const normalizedEntry: IocEntry = { ...entry, indicator };
    const idx = this.entries.length;
    const key = indicator.toLowerCase();

    switch (normalizedEntry.iocType) {
      case "sha256":
      case "sha1":
      case "md5": {
        const arr = this.hashIndex.get(key) ?? [];
        arr.push(idx);
        this.hashIndex.set(key, arr);
        break;
      }
      case "domain": {
        const arr = this.domainIndex.get(key) ?? [];
        arr.push(idx);
        this.domainIndex.set(key, arr);
        break;
      }
      case "ipv4":
      case "ipv6": {
        const arr = this.ipIndex.get(key) ?? [];
        arr.push(idx);
        this.ipIndex.set(key, arr);
        break;
      }
      case "url": {
        const arr = this.urlIndex.get(key) ?? [];
        arr.push(idx);
        this.urlIndex.set(key, arr);
        break;
      }
    }

    this.entries.push(normalizedEntry);
  }

  get size(): number {
    return this.entries.length;
  }

  get isEmpty(): boolean {
    return this.entries.length === 0;
  }

  merge(other: IocDatabase): void {
    for (const entry of other.entries) {
      this.addEntry(entry);
    }
  }

  // -- Loaders -------------------------------------------------------------

  /**
   * Load IOCs from a plain-text file (one indicator per line).
   * Empty lines and lines starting with '#' are skipped.
   */
  static async loadTextFile(path: string): Promise<IocDatabase> {
    const content = await fs.readFile(path, "utf-8");
    const db = new IocDatabase();
    for (const line of content.split("\n")) {
      const trimmed = line.trim();
      if (trimmed.length === 0 || trimmed.startsWith("#")) continue;
      const iocType = detectIocType(trimmed);
      if (iocType !== undefined) {
        db.addEntry({ indicator: trimmed, iocType });
      }
    }
    return db;
  }

  /**
   * Load IOCs from a CSV file.
   * Expected columns: indicator, type, description, source.
   */
  static async loadCsvFile(path: string): Promise<IocDatabase> {
    const content = await fs.readFile(path, "utf-8");
    const db = new IocDatabase();
    const lines = content.split("\n");

    let startIdx = 0;
    if (lines.length > 0) {
      const firstLower = lines[0].trim().toLowerCase();
      const isHeader =
        firstLower.startsWith("indicator,") ||
        firstLower.startsWith("indicator_type,") ||
        firstLower === "indicator";
      if (isHeader) {
        startIdx = 1;
      }
    }

    for (let i = startIdx; i < lines.length; i++) {
      if (lines[i].trim().length === 0) continue;
      const entry = parseCsvLine(lines[i]);
      if (entry) db.addEntry(entry);
    }

    return db;
  }

  /**
   * Load IOCs from a STIX 2.1 JSON bundle.
   */
  static async loadStixBundle(path: string): Promise<IocDatabase> {
    const content = await fs.readFile(path, "utf-8");
    let bundle: Record<string, unknown>;
    try {
      bundle = JSON.parse(content) as Record<string, unknown>;
    } catch (e) {
      throw new IocError(`STIX JSON parse error: ${e instanceof Error ? e.message : String(e)}`);
    }

    const db = new IocDatabase();
    const objects = bundle.objects;
    if (!Array.isArray(objects)) {
      throw new IocError("STIX bundle missing 'objects' array");
    }

    for (const obj of objects) {
      if (typeof obj !== "object" || obj === null) continue;
      const sdo = obj as Record<string, unknown>;
      if (sdo.type !== "indicator") continue;

      const pattern = sdo.pattern;
      if (typeof pattern !== "string") continue;

      const parsed = parseStixPattern(pattern);
      if (!parsed) continue;

      const description = typeof sdo.description === "string" ? sdo.description : undefined;
      const source = typeof sdo.name === "string" ? sdo.name : undefined;

      db.addEntry({
        indicator: parsed.indicator,
        iocType: parsed.iocType,
        description,
        source,
      });
    }

    return db;
  }

  // -- Matching ------------------------------------------------------------

  /**
   * Check a single timeline event against the IOC database.
   */
  matchEvent(event: TimelineEvent): IocMatch | undefined {
    const summary = event.summary.toLowerCase();
    const process = (event.process ?? "").toLowerCase();
    let raw = "";
    if (event.raw !== undefined) {
      try {
        raw = JSON.stringify(event.raw).toLowerCase();
      } catch {
        // Skip raw field if it cannot be serialized (e.g. circular refs).
      }
    }

    const allMatched: IocEntry[] = [];
    let matchField: string | undefined;

    // Helper to scan an index
    const scanIndex = (
      index: Map<string, number[]>,
      matcher: (haystack: string, needle: string) => boolean
    ): void => {
      for (const [needle, indices] of index) {
        const field = matcher(summary, needle)
          ? "summary"
          : matcher(process, needle)
            ? "process"
            : matcher(raw, needle)
              ? "raw"
              : undefined;
        if (field !== undefined) {
          if (matchField === undefined) matchField = field;
          for (const i of indices) {
            allMatched.push(this.entries[i]);
          }
        }
      }
    };

    // Hashes: plain substring match
    scanIndex(this.hashIndex, (haystack, needle) => haystack.includes(needle));

    // Domains: word-boundary match
    scanIndex(this.domainIndex, containsWordBounded);

    // IPs: word-boundary match
    scanIndex(this.ipIndex, containsWordBounded);

    // URLs: word-boundary match
    scanIndex(this.urlIndex, containsWordBounded);

    if (allMatched.length === 0) return undefined;

    return {
      event,
      matchedIocs: allMatched,
      matchField: matchField!,
    };
  }

  /**
   * Batch-match timeline events against the IOC database.
   */
  matchEvents(events: TimelineEvent[]): IocMatch[] {
    const matches: IocMatch[] = [];
    for (const event of events) {
      const m = this.matchEvent(event);
      if (m) matches.push(m);
    }
    return matches;
  }
}

File: packages/sdk/hush-py/src/clawdstrike/hunt/ioc.py
"""IOC (Indicator of Compromise) matching engine.

Port of ``hunt-correlate/src/ioc.rs``.
Supports loading from plain-text, CSV, and STIX 2.1 JSON bundles.
"""

from __future__ import annotations

import json
import re

from clawdstrike.hunt.errors import IocError
from clawdstrike.hunt.types import (
    IocEntry,
    IocMatch,
    IocType,
    TimelineEvent,
)

# ---------------------------------------------------------------------------
# Auto-detection
# ---------------------------------------------------------------------------

_HEX_RE = re.compile(r"^[0-9a-fA-F]+$")


def _is_ipv4(s: str) -> bool:
    parts = s.split(".")
    if len(parts) != 4:
        return False
    for p in parts:
        if not p:
            return False
        try:
            n = int(p)
        except ValueError:
            return False
        if n < 0 or n > 255:
            return False
        # Reject non-canonical representations (leading zeros, sign prefixes)
        if str(n) != p:
            return False
    return True


def detect_ioc_type(indicator: str) -> IocType | None:
    """Auto-detect the IOC type from the indicator string format.

    Returns ``None`` if the type cannot be determined.
    """
    trimmed = indicator.strip()
    if not trimmed:
        return None

    lower = trimmed.lower()

    # Hex hashes
    if len(lower) == 64 and _HEX_RE.match(lower):
        return IocType.SHA256
    if len(lower) == 40 and _HEX_RE.match(lower):
        return IocType.SHA1
    if len(lower) == 32 and _HEX_RE.match(lower):
        return IocType.MD5

    # URL
    if lower.startswith("http://") or lower.startswith("https://"):
        return IocType.URL

    # IPv4
    if _is_ipv4(trimmed):
        return IocType.IPV4

    # IPv6: contains colons and only hex digits / colons
    if ":" in trimmed and all(c in "0123456789abcdefABCDEF:" for c in trimmed):
        return IocType.IPV6

    # Domain: contains dot, no spaces, no slashes, no colons
    if "." in trimmed and " " not in trimmed and "/" not in trimmed and ":" not in trimmed:
        return IocType.DOMAIN

    return None


# ---------------------------------------------------------------------------
# Word-boundary matching
# ---------------------------------------------------------------------------


def _is_ioc_word_char(ch: str) -> bool:
    """IOC word character: alphanumeric, dot, or hyphen."""
    if len(ch) != 1:
        return False
    code = ord(ch)
    return (
        (ord("0") <= code <= ord("9"))
        or (ord("A") <= code <= ord("Z"))
        or (ord("a") <= code <= ord("z"))
        or ch == "."
        or ch == "-"
    )


def contains_word_bounded(haystack: str, needle: str) -> bool:
    """Check whether *needle* appears in *haystack* at word boundaries.

    Word characters for IOC matching are alphanumeric, ``'.'``, and ``'-'``.
    """
    if not needle:
        return False

    start = 0
    while True:
        pos = haystack.find(needle, start)
        if pos < 0:
            return False

        end_pos = pos + len(needle)

        left_ok = pos == 0 or not _is_ioc_word_char(haystack[pos - 1])
        right_ok = end_pos >= len(haystack) or not _is_ioc_word_char(haystack[end_pos])

        if left_ok and right_ok:
            return True

        # Advance past the first character of this match
        start = pos + 1


# ---------------------------------------------------------------------------
# CSV helpers
# ---------------------------------------------------------------------------


def _split_csv_fields(line: str) -> list[str]:
    """Split a CSV line handling double-quoted fields."""
    fields: list[str] = []
    current: list[str] = []
    in_quotes = False
    chars = iter(line)
    prev_quote = False

    for ch in chars:
        if in_quotes:
            if ch == '"':
                # Peek for escaped quote
                prev_quote = True
                in_quotes = False
            else:
                current.append(ch)
        elif prev_quote:
            prev_quote = False
            if ch == '"':
                current.append('"')
                in_quotes = True
            elif ch == ',':
                fields.append("".join(current).strip())
                current = []
            else:
                current.append(ch)
        elif ch == '"':
            in_quotes = True
        elif ch == ',':
            fields.append("".join(current).strip())
            current = []
        else:
            current.append(ch)

    if prev_quote:
        pass  # trailing quote already closed
    fields.append("".join(current).strip())
    return fields


_IOC_TYPE_STR_MAP: dict[str, IocType] = {
    "sha256": IocType.SHA256,
    "sha-256": IocType.SHA256,
    "sha1": IocType.SHA1,
    "sha-1": IocType.SHA1,
    "md5": IocType.MD5,
    "domain": IocType.DOMAIN,
    "domain-name": IocType.DOMAIN,
    "ipv4": IocType.IPV4,
    "ipv4-addr": IocType.IPV4,
    "ip": IocType.IPV4,
    "ipv6": IocType.IPV6,
    "ipv6-addr": IocType.IPV6,
    "url": IocType.URL,
}


def _parse_ioc_type_str(s: str) -> IocType | None:
    return _IOC_TYPE_STR_MAP.get(s.strip().lower())


def _parse_csv_line(line: str) -> IocEntry | None:
    fields = _split_csv_fields(line)
    if not fields:
        return None

    indicator = fields[0].strip()
    if not indicator:
        return None

    ioc_type: IocType | None = None
    if len(fields) > 1 and fields[1]:
        ioc_type = _parse_ioc_type_str(fields[1])
        if ioc_type is None:
            ioc_type = detect_ioc_type(indicator)
    else:
        ioc_type = detect_ioc_type(indicator)

    if ioc_type is None:
        return None

    description = fields[2] if len(fields) > 2 and fields[2] else None
    source = fields[3] if len(fields) > 3 and fields[3] else None

    return IocEntry(
        indicator=indicator,
        ioc_type=ioc_type,
        description=description,
        source=source,
    )


# ---------------------------------------------------------------------------
# STIX pattern parsing
# ---------------------------------------------------------------------------


def _stix_lhs_to_ioc_type(lhs: str) -> IocType | None:
    lower = lhs.lower()
    if "sha-256" in lower or "sha256" in lower:
        return IocType.SHA256
    if "sha-1" in lower or "sha1" in lower:
        return IocType.SHA1
    if "md5" in lower:
        return IocType.MD5
    if lower.startswith("domain-name"):
        return IocType.DOMAIN
    if lower.startswith("ipv4-addr"):
        return IocType.IPV4
    if lower.startswith("ipv6-addr"):
        return IocType.IPV6
    if lower.startswith("url"):
        return IocType.URL
    return None


def _parse_stix_pattern(pattern: str) -> tuple[str, IocType] | None:
    trimmed = pattern.strip()
    if not trimmed.startswith("[") or not trimmed.endswith("]"):
        return None

    inner = trimmed[1:-1]
    if "=" not in inner:
        return None

    lhs, rhs = inner.split("=", 1)
    lhs = lhs.strip()
    rhs = rhs.strip()

    if not rhs.startswith("'") or not rhs.endswith("'"):
        return None

    value = rhs[1:-1]
    if not value:
        return None

    ioc_type = _stix_lhs_to_ioc_type(lhs)
    if ioc_type is None:
        return None

    return (value, ioc_type)


# ---------------------------------------------------------------------------
# IocDatabase
# ---------------------------------------------------------------------------


class IocDatabase:
    """In-memory IOC database with indexed lookups."""

    def __init__(self) -> None:
        self._entries: list[IocEntry] = []
        self._hash_index: dict[str, list[IocEntry]] = {}
        self._domain_index: dict[str, list[IocEntry]] = {}
        self._ip_index: dict[str, list[IocEntry]] = {}
        self._url_index: dict[str, list[IocEntry]] = {}

    def add_entry(self, entry: IocEntry) -> None:
        """Add a single IOC entry and update indices."""
        indicator = entry.indicator.strip()
        if not indicator:
            return

        # Normalize indicator
        if indicator != entry.indicator:
            entry = IocEntry(
                indicator=indicator,
                ioc_type=entry.ioc_type,
                description=entry.description,
                source=entry.source,
            )

        self._entries.append(entry)
        key = indicator.lower()

        if entry.ioc_type in (IocType.SHA256, IocType.SHA1, IocType.MD5):
            self._hash_index.setdefault(key, []).append(entry)
        elif entry.ioc_type == IocType.DOMAIN:
            self._domain_index.setdefault(key, []).append(entry)
        elif entry.ioc_type in (IocType.IPV4, IocType.IPV6):
            self._ip_index.setdefault(key, []).append(entry)
        elif entry.ioc_type == IocType.URL:
            self._url_index.setdefault(key, []).append(entry)

    def __len__(self) -> int:
        return len(self._entries)

    def is_empty(self) -> bool:
        return len(self._entries) == 0

    def merge(self, other: IocDatabase) -> None:
        """Merge another database into this one."""
        for entry in other._entries:
            self.add_entry(entry)

    # -- Loaders -----------------------------------------------------------

    @classmethod
    def load_text_file(cls, path: str) -> IocDatabase:
        """Load IOCs from a plain-text file (one per line, ``#`` comments)."""
        try:
            with open(path) as f:
                content = f.read()
        except OSError as exc:
            raise IocError(f"failed to read {path}: {exc}") from exc

        db = cls()
        for line in content.splitlines():
            trimmed = line.strip()
            if not trimmed or trimmed.startswith("#"):
                continue
            ioc_type = detect_ioc_type(trimmed)
            if ioc_type is not None:
                db.add_entry(IocEntry(
                    indicator=trimmed,
                    ioc_type=ioc_type,
                    description=None,
                    source=None,
                ))
        return db

    @classmethod
    def load_csv_file(cls, path: str) -> IocDatabase:
        """Load IOCs from a CSV file.

        Expected columns: ``indicator, type, description, source``.
        """
        try:
            with open(path) as f:
                content = f.read()
        except OSError as exc:
            raise IocError(f"failed to read {path}: {exc}") from exc

        db = cls()
        lines = content.splitlines()
        if not lines:
            return db

        first = lines[0]
        first_lower = first.strip().lower()
        is_header = (
            first_lower.startswith("indicator,")
            or first_lower.startswith("indicator_type,")
            or first_lower == "indicator"
        )

        start_idx = 1 if is_header else 0
        for line in lines[start_idx:]:
            if not line.strip():
                continue
            entry = _parse_csv_line(line)
            if entry is not None:
                db.add_entry(entry)

        return db

    @classmethod
    def load_stix_bundle(cls, path: str) -> IocDatabase:
        """Load IOCs from a STIX 2.1 JSON bundle."""
        try:
            with open(path) as f:
                content = f.read()
        except OSError as exc:
            raise IocError(f"failed to read {path}: {exc}") from exc

        try:
            bundle = json.loads(content)
        except json.JSONDecodeError as exc:
            raise IocError(f"invalid JSON: {exc}") from exc

        objects = bundle.get("objects")
        if not isinstance(objects, list):
            raise IocError("STIX bundle missing 'objects' array")

        db = cls()
        for obj in objects:
            if not isinstance(obj, dict):
                continue

            sdo_type = obj.get("type", "")
            if sdo_type != "indicator":
                continue

            pattern = obj.get("pattern")
            if not isinstance(pattern, str):
                continue

            description = obj.get("description")
            source = obj.get("name")

            parsed = _parse_stix_pattern(pattern)
            if parsed is not None:
                indicator, ioc_type = parsed
                db.add_entry(IocEntry(
                    indicator=indicator,
                    ioc_type=ioc_type,
                    description=description,
                    source=source,
                ))

        return db

    # -- Matching ----------------------------------------------------------

    def match_event(self, event: TimelineEvent) -> IocMatch | None:
        """Match a single event against all IOCs. Returns accumulated matches or ``None``."""
        summary_lower = event.summary.lower()
        process_lower = (event.process or "").lower()
        raw_lower = ""
        if event.raw is not None:
            if isinstance(event.raw, str):
                raw_lower = event.raw.lower()
            else:
                try:
                    raw_lower = json.dumps(event.raw, default=str).lower()
                except (TypeError, ValueError, OverflowError, RecursionError):
                    # Ignore unserializable/circular payloads and continue
                    # scanning summary/process fields.
                    raw_lower = ""

        all_matched: list[IocEntry] = []
        match_field: str | None = None

        def _scan_index(
            index: dict[str, list[IocEntry]],
            matcher: object,
        ) -> None:
            nonlocal match_field
            for needle, entries in index.items():
                field = None
                if matcher(summary_lower, needle):  # type: ignore[operator]
                    field = "summary"
                elif matcher(process_lower, needle):  # type: ignore[operator]
                    field = "process"
                elif matcher(raw_lower, needle):  # type: ignore[operator]
                    field = "raw"
                if field is not None:
                    if match_field is None:
                        match_field = field
                    all_matched.extend(entries)

        # Hash index: plain substring match
        _scan_index(self._hash_index, lambda haystack, needle: needle in haystack)

        # Domain index: word-boundary match
        _scan_index(self._domain_index, contains_word_bounded)

        # IP index: word-boundary match
        _scan_index(self._ip_index, contains_word_bounded)

        # URL index: word-boundary match
        _scan_index(self._url_index, contains_word_bounded)

        if not all_matched:
            return None

        return IocMatch(
            event=event,
            matched_iocs=tuple(all_matched),
            match_field=match_field,  # type: ignore[arg-type]
        )

    def match_events(self, events: list[TimelineEvent]) -> list[IocMatch]:
        """Match multiple events. Returns all matches."""
        results: list[IocMatch] = []
        for event in events:
            m = self.match_event(event)
            if m is not None:
                results.append(m)
        return results


__all__ = [
    "detect_ioc_type",
    "contains_word_bounded",
    "IocDatabase",
]
